.github/workflows/pre-commit-autoupdate.yml

```diff
@@ -1,58 +1,69 @@
 name: Automatically update and run pre-commit hooks
 
-# Run daily at midnight.
 on:
+  # Run daily at midnight.
   schedule:
-    - cron: '*/5 * * * *'
+    - cron: '0 0 * * *'
+
+  # Also enable running manually.
+  workflow_dispatch:
 
 jobs:
-  # This workflow contains a single job called "autoupdate"
+  # This workflow contains a single job called 'autoupdate'.
   autoupdate:
     runs-on: ubuntu-latest
 
     steps:
-      - uses: actions/checkout@v2
+      - name: Check out the repository
+        # Don't use release tags for actions:  https://julienrenaux.fr/2019/12/20/github-actions-security-risk/
+        # Commit is equivalent to release tag v2.3.2.
+        uses: actions/checkout@2036a08e25fa78bbd946711a407b529a0a1204bf
 
       - name: Set the name of the branch for these changes
         id: branch
         run: echo "::set-output name=branchname::pre-commit-autoupdate"
 
       - name: Set up Python
-        uses: actions/setup-python@v2
+        # Commit is equivalent to release tag v2.1.2.
+        uses: actions/setup-python@24156c231c5e9d581bde27d0cdbb72715060ea51
         with:
-          python-version: >3
+          python-version: '>=3'
 
       - name: Install pre-commit
         run: pip install pre-commit
 
-      - name: ${{ github.action }}
+      - name: ${{ github.workflow }}
         run: |
           pre-commit autoupdate
-          pre-commit run --all-files
-
-      - name: Create a new branch pre-commit-autoupdate and push the changes to it
-        run: |
-          git checkout -B ${{ steps.branch.branchname }}
-          git add .
-          git commit -m "${{ github.action }}"
-          git push -fu origin ${{ steps.branch.branchname }}
+          until pre-commit run --all-files
+          do
+            printf "\nRe-running pre-commits...\n"
```

Comment:
On reflection this should probably read `echo -e "\e[33;1m Re-running pre-commits... \e[0m"`

Reply:
On the formatting :shrug: I just wanted an empty line above to mark a break between `pre-commit run` calls but I can highlight it instead.

On the loop:  If a hook makes changes, the exit code is non-zero, so the workflow is marked as failed.  We therefore need to catch that and run again until `pre-commit run` makes no changes and exits with `0`.  You have a point though, it may not terminate under certain circumstances, such as if two hooks make conflicting changes.  Not sure what the best approach is then.

Reply:
in that case maybe something like

```bash
pre-commit run --all-files || {
  git diff --quiet && {
    # pre-commit command returned an exit code, but no changes in the working directory
    echo command failed
    exit 1
  }
  while :; do
    echo some changes were made
    git add -A
    pre-commit run --all-files && break || {
      git diff --quiet && {
        echo No further changes were made
        break
      }
    }
  done
}
```

The above is agnostic about what pre-commit does. Istr though that pre-commit returns different exit codes depending on what happened.

Reply:
re. exit code: documentation search turns up empty. May have confused `pre-commit` with `pytest` here.

Reply:
Does this not do what we want?
```sh
while :; do
  pre-commit run --all-files && break || {
    git diff --quiet && {
      echo One or more pre-commit hooks failed and the repository contains no changes.
      exit 1
    }
  echo One or more pre-commit hooks failed and there are changes to the repository.
  echo -e "\e[1mRe-running pre-commit hooks...\e[0m"
  git add -A
  }
done
```

Reply:
It might well do, yes.